python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1110 @@
+        elif typ == CExpressionType_SCALAR:
+            self = ScalarExpression.__new__(ScalarExpression)
+        elif typ == CExpressionType_NOT:
+            self = NotExpression.__new__(NotExpression)
+        elif typ == CExpressionType_CAST:
+            self = CastExpression.__new__(CastExpression)
+        elif typ == CExpressionType_AND:
+            self = AndExpression.__new__(AndExpression)
+        elif typ == CExpressionType_OR:
+            self = OrExpression.__new__(OrExpression)
+        elif typ == CExpressionType_COMPARISON:
+            self = ComparisonExpression.__new__(ComparisonExpression)
+        elif typ == CExpressionType_IS_VALID:
+            self = IsValidExpression.__new__(IsValidExpression)
+        elif typ == CExpressionType_IN:
+            self = InExpression.__new__(InExpression)
+        # TODO(kszucs): implement it
+        # elif typ == CExpressionType_CUSTOM:
+        #     self = CustomExpression.__new__(CustomExpression)
+        else:
+            raise TypeError(typ)
+
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CExpression] unwrap(self):
+        return self.wrapped
+
+    def equals(self, Expression other):
+        return self.expression.Equals(other.unwrap())
+
+    def __str__(self):
+        return frombytes(self.expression.ToString())
+
+    def validate(self, Schema schema not None):
+        cdef:
+            shared_ptr[CSchema] sp_schema
+            CResult[shared_ptr[CDataType]] result
+        sp_schema = pyarrow_unwrap_schema(schema)
+        result = self.expression.Validate(deref(sp_schema))
+        return pyarrow_wrap_data_type(GetResultValue(result))
+
+    def assume(self, Expression given):
+        return Expression.wrap(self.expression.Assume(given.unwrap()))
+
+
+cdef class UnaryExpression(Expression):
+
+    cdef CUnaryExpression* unary
+
+    cdef void init(self, const shared_ptr[CExpression]& sp):
+        Expression.init(self, sp)
+        self.unary = <CUnaryExpression*> sp.get()
+
+
+cdef class BinaryExpression(Expression):
+
+    cdef CBinaryExpression* binary
+
+    cdef void init(self, const shared_ptr[CExpression]& sp):
+        Expression.init(self, sp)
+        self.binary = <CBinaryExpression*> sp.get()
+
+    @property
+    def left_operand(self):
+        return Expression.wrap(self.binary.left_operand())
+
+    @property
+    def right_operand(self):
+        return Expression.wrap(self.binary.right_operand())
+
+
+cdef class ScalarExpression(Expression):
+
+    cdef CScalarExpression* scalar
+
+    def __init__(self, value):
+        cdef:
+            shared_ptr[CScalar] scalar
+            shared_ptr[CScalarExpression] expression
+
+        if isinstance(value, bool):
+            scalar = MakeScalar(<c_bool>value)
+        elif isinstance(value, float):
+            scalar = MakeScalar(<double>value)
+        elif isinstance(value, int):
+            scalar = MakeScalar(<int64_t>value)
```

Comment:
As a temporary measure until Python has full scalar support, it might be useful to wrap `Scalar::Parse` here as `ScalarExpression.Parse` or so. That would allow creation of `ScalarExpression`s of most types from string representations

Reply:
I'd still need to pass a datatype, so it'd require to infer the expected type on python side. I already have an ongoing patch for the scalar support, which I'm going to keep working on once we merge this patch.